### DISCUSSION TITLE
Avoiding Duplicate Updates with RavenDB Changes API in Kubernetes

### DISCUSSION THREAD

**Discussion by andersgranberg** [ravendb/ravendb#19517] [Q&A] [answered]
Hello

I'm setting up a RavenDB Changes API listener hosted within a Kubernetes cluster. My application runs across three pods, each maintaining an identical setup, and I want to avoid duplicated database updates from multiple listeners triggered by the same change.

I need a robust solution for ensuring that only one instance (or pod) performs the update on any given change. My setup also needs to handle scenarios where a pod failure prompts a new leader to take over without service disruption.

Do you have any best practice recommendations for implementing this? Would leader election or distributed locking be the preferred approach for minimizing duplicate updates within a RavenDB environment? And are there any specific considerations for configuring the Changes API with this setup in mind?

Thank you for any guidance you can provide!

**Comment by lukaszdobrzynski** [comment, marked as answer]
What you're probably looking for are `RavenDB Subscriptions` with `TakeOver` strategy option applied. The strategy ensures that the subscription will always be held by the last connected client application thus allowing your system to avoid duplicate processing and updates. To find out more about the setup and available options I strongly recommend that you read [this chapter](https://ravendb.net/learn/inside-ravendb-book/reader/4.0/5-batch-processing-with-subscriptions#batch-processing-with-subscriptions) from Oren's book or have a look at the [documentation](https://ravendb.net/docs/article-page/6.2/csharp/client-api/data-subscriptions/what-are-data-subscriptions).
Another important thing to mention about subscriptions is that document processing is always guaranteed on the client side. This means that the server persists the processing progress, allowing you to pause and continue the processing. This is something you cannot expect with Changes API.    
Also, please remember that subscriptions are cluster operations and they require your cluster to be fully operational.

**Reply by andersgranberg** [reply]
Thanks for the reply. We're also using subscriptions for other purposes and this works well as you describe cluster wide. The main reason for also using Changes API is to monitor Counter changes when a value is incremented. Once a counter is incremented a series of other documents will be created (for instance notifications and activity logs). This can't, as far as I understand, not be done with subscriptions?

**Reply by lukaszdobrzynski** [reply]
RavenDB does not have built-in support for leader election between multiple clients to determine which one should actively listen to the Changes API. This is another crucial difference between Changes API and Subscriptions. 

However, RavenDB does support subscriptions on document counters, which allows you to react to counter changes directly within the subscription mechanism ([documentation link](https://ravendb.net/docs/article-page/6.2/csharp/client-api/data-subscriptions/creation/examples#create-subscription---include-counters)).
Just load a document, increment a counter, mark the document as changed and send the data to the server as part of a single transaction. If you include your counters as part of the subscription building process, the connected client will be able to access the counter values and respond accordingly (create a document,  a notification, a log activity, etc.).
